### PR TITLE
Refactor pj_transform, reintroduce support for vertical scaling

### LINKED ISSUE #833: cs2cs 5.0.0 ignores +vunits
As reported in [Debian Bug #889936](https://bugs.debian.org/889936), survex 1.2.32 failed to build due to test failures.

The maintainer (@ojwb) responded that its caused by +vunits being ignored:
> On Thu, Feb 08, 2018 at 10:55:00PM +0100, Bas Couwenberg wrote:
> > Your package FTBFS due to missing compatibility with Proj 5.0.0:
> > ```
> >  ./csbadsdfix.svx:2: error: Station "1" fixed before CS command first used
> >  ./csbadsdfix.svx:3:5: error: Unknown coordinate system
> >   *cs EPSG:-1
> >       ^~~~~~~
> >  ./csbadsdfix.svx:4:5: error: Unknown coordinate system
> >   *cs ERSI:1234
> >       ^~~~~~~~~
> >  ./csbadsdfix.svx:5:5: error: Unknown coordinate system
> >   *cs EUR79Z31
> >       ^~~~~~~~
> > ```
> [...]
> 
> This isn't the problem (these testcases are testing that we reject
> coordinate systems which aren't valid, so these errors are expected and
> correct).
> 
> The test failure seems to be actually due to the new PROJ4 ignoring
> `+vunits` entirely and can be reproduced without involving Survex by
> using `cs2cs`:
> ```
> $ echo 36000 83000 5250|cs2cs +proj=tmerc +lat_0=0 +lon_0=13d20 +k=1 +x_0=0 +y_0=-5200000 +ellps=bessel +towgs84=577.326,90.129,463.919,5.137,1.474,5.297,2.4232 +vunits=ft +to +proj=tmerc +lat_0=0 +lon_0=13d20 +k=1 +x_0=0 +y_0=-5200000 +ellps=bessel +towgs84=577.326,90.129,463.919,5.137,1.474,5.297,2.4232
> 36000.00	83000.00 5250.00
> ```
> In older versions this converted the altitude from feet to metres, and
> the output was:
> ```
> 36000.00	83000.00 1600.20
> ```

Debug output of `cs2cs` 4.9.3:
```
$ echo 36000 83000 5250 | PROJ_DEBUG=3 cs2cs +proj=tmerc +lat_0=0 +lon_0=13d20 +k=1 +x_0=0 +y_0=-5200000 +ellps=bessel +towgs84=577.326,90.129,463.919,5.137,1.474,5.297,2.4232 +vunits=ft +to +proj=tmerc +lat_0=0 +lon_0=13d20 +k=1 +x_0=0 +y_0=-5200000 +ellps=bessel +towgs84=577.326,90.129,463.919,5.137,1.474,5.297,2.4232
pj_open_lib(proj_def.dat): call fopen(/usr/share/proj/proj_def.dat) - succeeded

pj_open_lib(proj_def.dat): call fopen(/usr/share/proj/proj_def.dat) - succeeded

36000.00        83000.00 1600.20
```

Debug output of `cs2cs` 5.0.0:
```
$ echo 36000 83000 5250 | PROJ_DEBUG=3 cs2cs +proj=tmerc +lat_0=0 +lon_0=13d20 +k=1 +x_0=0 +y_0=-5200000 +ellps=bessel +towgs84=577.326,90.129,463.919,5.137,1.474,5.297,2.4232 +vunits=ft +to +proj=tmerc +lat_0=0 +lon_0=13d20 +k=1 +x_0=0 +y_0=-5200000 +ellps=bessel +towgs84=577.326,90.129,463.919,5.137,1.474,5.297,2.4232
get_init: searching cache for key: [proj_def.dat:general]

get_init: searching on in init files for [proj_def.dat:general]

get_init_string: searching for section [general] in init file [proj_def.dat]

pj_open_lib(proj_def.dat): call fopen(/usr/share/proj/proj_def.dat) - succeeded

key=proj_def.dat:general, value: [ellps=WGS84]

get_init: got [ellps=WGS84], paralist[0,1]: [ellps=WGS84,(empty)]

get_init: searching cache for key: [proj_def.dat:tmerc]

get_init: searching on in init files for [proj_def.dat:tmerc]

get_init_string: searching for section [tmerc] in init file [proj_def.dat]

pj_open_lib(proj_def.dat): call fopen(/usr/share/proj/proj_def.dat) - succeeded

pj_ellipsoid - final: a=6377397.155 f=1/299.153, errno=0
pj_ellipsoid - final:    ellps=bessel
get_init: searching cache for key: [proj_def.dat:general]

get_init: searching cache for key: [proj_def.dat:tmerc]

get_init: searching on in init files for [proj_def.dat:tmerc]

get_init_string: searching for section [tmerc] in init file [proj_def.dat]

pj_open_lib(proj_def.dat): call fopen(/usr/share/proj/proj_def.dat) - succeeded

pj_ellipsoid - final: a=6377397.155 f=1/299.153, errno=0
pj_ellipsoid - final:    ellps=bessel
36000.00        83000.00 5250.00
```

### PR DESCRIPTION
**Pull request by busstoptaktik**
Resolves #833

### CODE REVIEW

**Comment by sebastic** [comment]
Why the refactoring?

The changes look a little invasive and may cause different regressions.

**Comment by busstoptaktik** [comment]
> Why the refactoring? The changes look a little invasive and may cause different regressions.

Actually they are not anywhere as invasive as the diff makes it look. 

Basically it is just a matter of moving, block-by-block, branches of the massive, `if {...} else if {if...else} else {...}` hierarchy out to individual functions.

I did it primarily to make sure I placed the vertical scaling in a place where it did not accidentally break the flow by e.g. hijacking an `else`-branch: The huge block structure was easy to get lost in.

Secondarily, I did it to do a double check on the recently introduced (by-and-large) reimplementation of the same functionality in `pj_fwd.c/pj_inv.c`.

Tertiarily, once getting going, I saw that the changes were few, cut-and-paste style, easy to implement, had fine test coverage [1], *and gave a much better general view of what actually goes on in `pj_transform()`, hence providing a good exposition of "how geodesy works".*

I refrained from merging the `projected_to_geographic` and `geographic_to_projected` into one function, since that would have required actual code changes, risking the kind of accidental breakage, you mention.

So, to summarize: The refactoring looks extensive, but really isn't. It reduced the risk of introducing new bugs when re-introducing the vertical scaling, and it gave me a chance of cross checking the code in `pj_fwd.c/pj_inv.c`. I stopped the refactoring at the step before it required major changes of the flow logic. This will be a task for 6.0 or 7.0 if `pj_transform()` is still needed at that point in time.

[1] Test coverage is only 61%, but the branches not hit during test is basically the error handlers, which is a known limitation in the existing test setup. We're improving that in `gie`, where we have recently focused on testing whether failing cases fails in the expected way.

**Comment by sebastic** [comment]
That sounds sane, the biggest issue is that the refactoring is not easy to review because it's all in a single commit.

I've cherry-picked the changes from this PR and included them as a patch in the proj Debian package, and confirmed that it fixes the regression causing the survex build (and tests) to fail.

**Comment by sebastic** [comment]
This change seems to cause a regression in [shapelib](http://shapelib.maptools.org/) `shpproj`:
```
make[4]: Entering directory '/build/shapelib/contrib'
FAIL: tests/shpproj.sh
============================================
   shapelib 1.4.1: contrib/test-suite.log
============================================

# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: tests/shpproj.sh
======================

./tests/shpproj.sh: line 21:  7390 Segmentation fault      $top_builddir/contrib/shpproj "$testdir/test" "$testdir/test_1" -i=geographic -o="init=nad83:1002 units=us-ft"

using default file proj params from <- /build/shapelib/contrib/tests/test_2.prj

******* Test Failed *********
1c1
< Shapefile Type: Point   # of Shapes: 5
---
> Shapefile Type: Point   # of Shapes: 0
3,29c3,4
< File Bounds: (-84.15596,34.54927,0,0)
<          to  (-83.521622,34.992767,0,0)
< 
< Shape:0 (Point)  nVertices=1, nParts=0
<   Bounds:(-83.5495,34.992401, 0)
<       to (-83.5495,34.992401, 0)
<      (-83.5495,34.992401, 0)  
< 
< Shape:1 (Point)  nVertices=1, nParts=0
<   Bounds:(-83.521622,34.992767, 0)
<       to (-83.521622,34.992767, 0)
<      (-83.521622,34.992767, 0)  
< 
< Shape:2 (Point)  nVertices=1, nParts=0
<   Bounds:(-84.016815,34.67276, 0)
<       to (-84.016815,34.67276, 0)
<      (-84.016815,34.67276, 0)  
< 
< Shape:3 (Point)  nVertices=1, nParts=0
<   Bounds:(-84.15596,34.648624, 0)
<       to (-84.15596,34.648624, 0)
<      (-84.15596,34.648624, 0)  
< 
< Shape:4 (Point)  nVertices=1, nParts=0
<   Bounds:(-83.619515,34.54927, 0)
<       to (-83.619515,34.54927, 0)
<      (-83.619515,34.54927, 0)  
---
> File Bounds: (0,0,0,0)
>          to  (0,0,0,0)
FAIL tests/shpproj.sh (exit status: 1)

============================================================================
Testsuite summary for shapelib 1.4.1
============================================================================
# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See contrib/test-suite.log
============================================================================
```

**gdb backtrace**:
```gdb
(gdb) run
Starting program: /build/shapelib/contrib/.libs/shpproj ./contrib/tests/test ./contrib/tests/test_1 -i=geographic -o=init=nad83:1002\ units=us-ft
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
pj_transform (srcdefn=srcdefn@entry=0x55555575d2c0, dstdefn=dstdefn@entry=0x55555575ec20, point_count=<optimized out>, point_offset=1, point_offset@entry=0, x=0x555555762f20, y=<optimized out>, z=0x0) at pj_transform.c:516
516     pj_transform.c: No such file or directory.
(gdb) bt full
#0  pj_transform (srcdefn=srcdefn@entry=0x55555575d2c0, dstdefn=dstdefn@entry=0x55555575ec20, point_count=<optimized out>, point_offset=1, point_offset@entry=0, x=0x555555762f20, y=<optimized out>, z=0x0) at pj_transform.c:516
        err = <optimized out>
#1  0x0000555555555e34 in SHPProject (psCShape=0x555555762e80, inproj=0x55555575d2c0, outproj=0x55555575ec20) at shpgeo.c:206
        j = <optimized out>
#2  0x00005555555556a1 in main (argc=<optimized out>, argv=<optimized out>) at shpproj.c:227
        old_SHP = 0x55555575b260
        new_SHP = 0x55555575f3f0
        old_DBF = 0x55555575c8a0
        new_DBF = 0x555555761910
        nShapeType = 1
        nEntities = 5
        i = 0
        j = <optimized out>
        psCShape = 0x555555762e80
        ifp = <optimized out>
        prjFileName = '\000' <repeats 119 times>
        parg = "units=us-ft\000\060\060\062\000\300\342\377\377\377\177\000\000\320\342\377\377\377\177\000\000\230\352\377\367\377\177", '\000' <repeats 26 times>, "\377\377\377\377", '\000' <repeats 11 times>
        inarg = <optimized out>
        outarg = <optimized out>
        DBFRow = 0x555555762e40 ""
        in_args = {0x55555575ce60 "geographic", 0x7fffffffe208 "\240\316uUUU", 0x7fffffffe204 "UU", 0x0, 0x0, 0x7ffff7ffa298 "", 0x7ffff7ffa1a8 "", 0x7ffff737ed07 "__vdso_getcpu", 
          0x6562b026 <error: Cannot access memory at address 0x6562b026>, 0x1958ac0 <error: Cannot access memory at address 0x1958ac0>, 0x7fffffffe370 "`\344\377\377\377\177", 0x7fffffffe2d0 "", 0x7fffffffe2c0 "\377\377\377\377", 
          0x7fffffffe208 "\240\316uUUU", 0x7ffff7ffe738 "\020\347\377\367\377\177", 0x1 <error: Cannot access memory at address 0x1>}
        out_args = {0x55555575ce80 "init=nad83:1002", 0x55555575cea0 "units=us-ft", 0x2e322e <error: Cannot access memory at address 0x2e322e>, 0x6562b026 <error: Cannot access memory at address 0x6562b026>, 
          0x7ffff7ffea98 "\310\351\377\367\377\177", 0x7fffffffe368 "\220[UUUU", 0x7fffffffe3a0 "", 0x7ffff7ffe710 "", 0x0, 0x7ffff7de2017 "H\203\304\060\205\300~\251I\213\f$L\211\375H\203|$P", 0x0, 0x7fffffffe3a0 "", 0x0, 0x0, 0x0, 
          0x7ffff7ffe710 ""}
        in_argc = <optimized out>
        out_argc = 2
        arglst = <optimized out>
        orig_prj = 0x55555575d2c0
        new_prj = 0x55555575ec20
```

**Comment by sebastic** [comment]
Reverting the changes from this PR and only keeping f41fad3ac2bff401456f31dd3273e163ea7d09af on top of 5.0.0 fixes the shapelib testsuite failure, the cause is definitely in the changes for the PR.

File: src/pj_transform.c [hunk 1]
```diff
@@ -81,388 +93,468 @@ static const int transient_error[60] = {
...
-                {
-                    x[point_offset*i] *= dstdefn->fr_meter;
-                    y[point_offset*i] *= dstdefn->fr_meter;
-                    z[point_offset*i] *= srcdefn->fr_meter;
-                }
-            }
-        }
-    }
+    /* Nothing to do? */
+    if (P->vto_meter==0.0)
```
Comment (sebastic) [inline src/pj_transform.c:400]: Shouldn't this be:
```
if (P->vto_meter==1.0)
```

In 4.9.3 the code was:
```c
/* -------------------------------------------------------------------- */
/*      Transform Z to meters if it isn't already.                      */
/* -------------------------------------------------------------------- */
    if( srcdefn->vto_meter != 1.0 && z != NULL )
    {
        for( i = 0; i < point_count; i++ )
            z[point_offset*i] *= srcdefn->vto_meter;
    }
```

There is also no test for `z != NULL`.

**Comment by sebastic** [comment]
This patch fixes the shapelib regression:
```patch
Description: Fix height_unit() in proj_transform.c.
Author: Bas Couwenberg <sebastic@debian.org>

--- a/src/pj_transform.c
+++ b/src/pj_transform.c
@@ -397,7 +397,10 @@ static int height_unit (PJ *P, PJ_DIRECT
         fac = P->vfr_meter;
 
     /* Nothing to do? */
-    if (P->vto_meter==0.0)
+    if (fac == 1.0)
+        return 0;
+
+    if (z == NULL)
         return 0;
 
     for (i = 0;  i < n;  i++)
```

It implements the same checks as in proj 4.9.3.

File: src/pj_transform.c (hunk 1 above)
Comment (busstoptaktik) [reply src/pj_transform.c:400]: @sebastic: That's certainly a blunder - I'll clean this up immediately. Sorry for merging this before seeing your review. I'll open another PR cleaning this up

**Comment by busstoptaktik** [comment]
Follow-up in #848, implementing the fix from @sebastic

**Comment by kbevers** [comment]
cherry-picked into 5.0 branch
